rfcs/0124-worker-manager.md

```diff
@@ -0,0 +1,271 @@
+cloud provider concepts.
+
+The worker manager will have bidding strategies and providers.  The bidding
+strategies will provide different ways to decide whether new capacity is
+needed.  Providers will understand how to suggest bids for a given capacity
+requirement as well as allocating the selected bids.
+
+The bidding strategy will be configurable in the worker configuration by
+setting the identifier of the strategy for that worker configuration.
+Providers will be configured globally.  Each worker configuration will need to
+opt in to each provider as well as give provider specific configuration.
+
+### Bidding
+Bidding strategies will take as inputs the current worker type state, the
+recent history of a worker type, the worker configuration and the amount of
+work pending from the Queue.  They will calculate the number of capacity units
+which are needed.  If that number is greater than zero, they will request bids
+from the providers.
+
+The bid request will be the number of capacity units needed.  The bid response
+will be a list of bids from each configured provider.  These bids will have
+standard response entries as well as an opaque data field which the provider
+will use the complete the bid if it is selected.  The standard response entries
+will be things like the amount of capacity for that bid, a price per capacity
+unit in the bid and a rating of how reliable the provider thinks the bid will
+be.  The bidding strategy will pass the selected bid to the provider for
+allocation.  The default bidding strategy will match the current
+aws-provisioner algorithm.
+
+In order to support development of experimental bidding strategies, a bidding
+strategy which can forward each request to an external service will be
+provided.  Providers do not need to be implemented directly inside of the
+worker manager, but can instead act as a bridge to an external service.
+
+## Credentials
+Taskcluster credentials will be given out by the worker manager.  These
+credentials will be used to obtain trusted credentials onto workers which boot
+from worker images which have no embedded credentials.  Each cloud or
+datacenter will have its own unique requirements and the specifics should be
+implemented in the providers.
+
+One special case is machines which are self-hosted in data centers.  These
+machines do not have any management APIs from which to derive credentials, so a
+different approach is needed.  A small server will run in the data center which
+will expose an API to workers in the data center.  It will determine the IP
+address of the requester, then do a forward and reverse dns lookup to determine
+the host name of the machine.  The server will then request credentials from the
+main worker manager through a Taskcluster-scope protected API and forward them
+onto the worker.
+
+# Open Questions
+- Will we deprecate provisionerId? -- separate RFC?
+- Email Provisioning strategy -- Email a mailing list when a worker type needs
```

Comment:
This could just a "notify provisioner strategy" where you provide whatever inputs taskcluster-notify takes. You would get email/irc/whatever from that.

Reply:
Sounds like a better idea than reimplementing emailing.